rfc/0068-kv-error-map-v2.md

```diff
@@ -0,0 +1,280 @@
+
+Attributes are the most important part of the error map as they convey how an error code should be handled. An error code may have more than a single attribute.
+
+- `item-only`: This attribute means that the error is related to a constraint failure regarding the item itself, i.e. the item does not exist, already exists, or its current value makes the current operation impossible. Retrying the operation when the item's value or status has changed may succeed.
+- `invalid-input`: This attribute means that a user's input was invalid because it violates the semantics of the operation, or exceeds some predefined limit.
+- `fetch-config`: The client's cluster map may be outdated and requires updating. The client should obtain a newer configuration.
+- `conn-state-invalidated`: The current connection is no longer valid. The client must reconnect to the server. Note that the presence of other attributes may indicate an alternate remedy to fixing the connection without a disconnect, but without special remedial action a disconnect is needed.
+- `auth`: The operation failed because the client failed to authenticate or is not authorized to perform this operation. Note that this error in itself does not mean the connection is invalid, unless conn-state-invalidated is also present.
+- `special-handling`: This error code must be handled specially. If it is not handled, the connection must be dropped.
+- `support`: The operation is not supported, possibly because the of server version, bucket type, or current user.
+- `temp`: This error is transient. Note that this does not mean the error is retriable.
+- `internal`: This is an internal error in the server.
+- `retry-now`: The operation may be retried immediately.
+- `retry-later`: The operation may be retried after some time.
+- `subdoc`: The error is related to the subdocument subsystem.
+- `dcp`: The error is related to the DCP subsystem.
+- `rate-limit`: The error is related to rate limiting being applied to the current connection or user.
+
+The most up-to-date list of attributes can to be found under https://github.com/couchbase/memcached/blob/master/docs/ErrorMap.md#error-attributes.
+
+It is important to note that a single error code may contain more than one attribute. Some attributes are informational by nature, while others may affect the flow of the SDK, and yet others may affect the flow of the application. Additionally, to support clean forwards compatibility, the SDKs must be designed such that unrecognized attributes are silently ignored and behaviours defined for any of the known attributes are still applied.
+
+It is important to reiterate while the attributes are important for unknown error codes to decide what to do next, the SDKs must favor backwards compatibility for known status codes / responses, and the error map attributes should only be used to decide in the case that an unknown status code is encountered.
+
+# Retry Intervals
```

Comment:
Why do we have this section in the V2 RFC? I thought we decided to ingore it completely and don't allow the server to enforce/change retry behaviour.

Reply:
.NET ignores this presently.

Reply:
According to the currently accepted RFC, we should only look if it is retryable or not -- and not look at the type of retry delays etc. https://github.com/couchbaselabs/sdk-rfcs/blob/master/rfc/0049-sdk3-retry-handling.md#kv-error-map-handling